### DISCUSSION TITLE
Excessive compile time

### DISCUSSION THREAD

**Discussion by cancech** [micronaut-projects/micronaut-core#6937] [Q&A] [answered]
I recall seeing in the Micronaut documentation that compiling a Micronaut application takes longer (due to the need generate the binding code), but is there a measure for how much extra time this should take?

I have been working to migrate our in-house development libraries/framework from Spring/Dagger to Micronaut, and now that I'm reaching the end of this journey I'm noticing just how much longer the compilation takes. If I'm compiling after a simple code change it's not so bad, however if I find myself compiling everything (such as after updating a gradle file or clean) then it takes an excessive amount of time. Due to changes in a gradle file what I presume to be a full build was kicked off, and in the extra mount of time required is just staggering. I've tried this with two different items and the jump is from less than 2 mins (~100 secs originally with Spring/Dagger) to 3+ hours with Micronaut. We're using nested gradle projects, where a common "plugin" is used to handle the Micronaut aspect

```
plugins {
    id "io.micronaut.library"
}

project.ext {
    picocliVersion = '4.6.2'
    micronautVersion = '3.2.1'
    micronautTestVersion = '3.0.5'
}

dependencies {
    annotationProcessor "info.picocli:picocli-codegen:$picocliVersion"
    annotationProcessor "io.micronaut:micronaut-validation:$micronautVersion"
    annotationProcessor "io.micronaut:micronaut-inject-java:$micronautVersion"
    
    implementation "info.picocli:picocli:$picocliVersion"
    implementation "io.micronaut.picocli:micronaut-picocli:4.0.1"
    implementation "io.micronaut:micronaut-cli:1.3.7"
    implementation "io.micronaut:micronaut-runtime:$micronautVersion"
    implementation "io.micronaut:micronaut-inject:$micronautVersion"
    implementation "io.micronaut:micronaut-validation:$micronautVersion"
    implementation "io.micronaut:micronaut-core:$micronautVersion"
    implementation "io.micronaut:micronaut-websocket:$micronautVersion"
    
    testAnnotationProcessor "io.micronaut:micronaut-inject-java:$micronautVersion"
    testImplementation "io.micronaut.test:micronaut-test-core:$micronautTestVersion"
    testImplementation "io.micronaut.test:micronaut-test-junit5:$micronautTestVersion"
}

tasks.withType(JavaCompile) {
    options.encoding = 'UTF-8'
    options.compilerArgs = [
        '-Amicronaut.processing.incremental=true',
        '-Amicronaut.processing.annotations=com.lmcda.vista.*',
        '-parameters'
    ]
}
```

I'm performing the build/launch from within Eclipse using the gradle tasks. Is this expected behavior? Is there something incorrect in the above gradle configuration? Is there something that I can do to optimize or speed things up?

Thanks!

**Comment by graemerocher** [comment, marked as answer]
3 hours doesn’t make sense probably you need to assign more memory to the compiler

**Reply by cancech** [reply]
I gave the compiler as much memory as I could and that drastically dropped the compile time. So what pre-Micronaut was taking ~100 seconds is taking ~140 with Micronaut, and another item which was taking ~30 mins to compile pre-Micronaut is taking ~45 mins with Micronaut. Is this more in the realm of expectation?

And more generally, can you also confirm that the gradle config I posted is reasonable? I have three annotation processors (picocli, micronaut-validation, and micronaut-inject-java) and just want to make sure that having these three configured doesn't cause any issues.

Thanks!

**Reply by graemerocher** [reply]
Yes a certain amount of overhead is expected. 30 minute compile times are a lot though, perhaps that project needs breaking up :)

**Reply by cancech** [reply]
Thanks for confirming!
